src/v3/src/hooks/useOnSubmit.ts

```diff
@@ -140,36 +174,73 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
       if (!newTransaction.nextStep && newTransaction.availableSteps?.length) {
         [newTransaction.nextStep] = newTransaction.availableSteps;
       }
-      setIdxTransaction(newTransaction);
       const transactionHasWarning = (newTransaction.messages || []).some(
         (message) => message.class === MessageType.WARNING.toString(),
       );
-      if (newTransaction.requestDidSucceed === false) {
-        events?.afterError?.(
-          getEventContext(newTransaction),
-          getErrorEventContext(newTransaction.rawIdxState),
-        );
-      }
       const isClientTransaction = !newTransaction.requestDidSucceed
         || (areTransactionsEqual(currTransaction, newTransaction) && transactionHasWarning);
-      setIsClientTransaction(isClientTransaction);
-      if (isClientTransaction && !newTransaction.messages?.length) {
-        setMessage({
-          message: loc('oform.errorbanner.title', 'login'),
-          class: 'ERROR',
-          i18n: { key: 'oform.errorbanner.title' },
-        } as IdxMessage);
+
+      const onSuccess = (resolve?: (val: unknown) => void) => {
+        setIdxTransaction(newTransaction);
+        if (newTransaction.requestDidSucceed === false) {
+          events?.afterError?.(
+            getEventContext(newTransaction),
+            getErrorEventContext(newTransaction.rawIdxState),
+          );
+        }
+        setIsClientTransaction(isClientTransaction);
+        if (isClientTransaction && !newTransaction.messages?.length) {
+          setMessage({
+            message: loc('oform.errorbanner.title', 'login'),
+            class: 'ERROR',
+            i18n: { key: 'oform.errorbanner.title' },
+          } as IdxMessage);
+        }
+        setStepToRender(stepToRender);
+        setLoading(false);
+        resolve?.(true);
+      };
```

Comment:
Needed a re-usable way to execute this group of functions (between the reg callback and in non registration flows/scenarios)